src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/AvoidUninstantiatedInternalClasses.cs

```diff
@@ -104,16 +107,47 @@ public override void Initialize(AnalysisContext analysisContext)
 
-                    // Consider class type declared in the CoClass attribute as instantiated
-                    if (coClassAttributeSymbol != null &&
-                        type.TypeKind == TypeKind.Interface &&
-                        type.GetAttributes().FirstOrDefault(x => x.AttributeClass.Equals(coClassAttributeSymbol)) is AttributeData coClassAttribute &&
-                        coClassAttribute.ConstructorArguments.Length == 1 &&
-                        coClassAttribute.ConstructorArguments[0].Kind == TypedConstantKind.Type &&
-                        coClassAttribute.ConstructorArguments[0].Value is INamedTypeSymbol typeSymbol &&
-                        typeSymbol.TypeKind == TypeKind.Class)
+                    // Consider class types declared in the CoClassAttribute or DesignerAttribute as instantiated
+                    if ((coClassAttributeSymbol != null || designerAttributeSymbol != null) &&
+                        (type.TypeKind == TypeKind.Class || type.TypeKind == TypeKind.Interface))
                     {
-                        instantiatedTypes.TryAdd(typeSymbol, null);
+                        bool isCoClassHandled = false;
+                        foreach (var attribute in type.GetAttributes())
+                        {
+                            if (coClassAttributeSymbol != null &&
+                                !isCoClassHandled &&
+                                attribute.AttributeClass.Equals(coClassAttributeSymbol))
+                            {
+                                isCoClassHandled = true;
+                                if (attribute.ConstructorArguments.Length == 1 &&
+                                    attribute.ConstructorArguments[0].Kind == TypedConstantKind.Type &&
+                                    attribute.ConstructorArguments[0].Value is INamedTypeSymbol typeSymbol &&
+                                    typeSymbol.TypeKind == TypeKind.Class)
+                                {
+                                    instantiatedTypes.TryAdd(typeSymbol, null);
+                                }
+                            }
+
+                            if (designerAttributeSymbol != null &&
+                                attribute.AttributeClass.Equals(designerAttributeSymbol))
+                            {
+                                switch (attribute.ConstructorArguments[0].Value)
+                                {
+                                    case string designerTypeName:
+                                        var nameParts = designerTypeName.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
+                                        if (nameParts.Length >= 2 &&
+                                            nameParts[1].Trim().Equals(context.Compilation.AssemblyName, StringComparison.Ordinal))
+                                        {
+                                            instantiatedTypeNames.TryAdd(nameParts[0].Trim(), null);
```

Comment:
I would avoid tracking names. Can we instead try to invoke compilation.GetTypeByMetadataName for the name and see if it is a non-null symbol whose containing assembly matches compilation's assembly.

Reply:
Indeed that's a way better (faster and more robust) solution... I don't know why I didn't think of it.